### PR TITLE
Add ivalue::type(), part 1

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26271 [JIT] add type refinements for isinstance checks
* #26270 [jit] refactor extra sugared values
* #26269 [jit] add dynamic isinstance
* #26268 Allow types as node attributes
* #26145 [JIT] add CondValue to unify refinements and code emission
* #26061 [jit] Refactor emitIsInstance
* #25447 Use static type information to restore type tags
* **#25439 Add ivalue::type(), part 1**

This introduces a type() method on IValue that returns the tagged type
of the IValue. The intention is that this value is always present/accurate,
making it possible for clients to recover the Type from an IValue.
Currently our APIs here are incomplete: they can sometimes recover a type but not always.

This PR adds the function, and cleans up remaining cases where Lists/Dicts are not
tagged. However, this information does not survive serialization unchanged.

A second PR will use the type information in the ClassType being serialized
to fixup the serialized ivalues to have the correct types again.
After this patch it will be save to remove our incomplete APIs for recovering types.

Differential Revision: [D17125595](https://our.internmc.facebook.com/intern/diff/D17125595)

### CODE REVIEW

**Comment by eellison** [comment]
How do you see this interacting with users who have to construct IValues in c++ ?

**Comment by zdevito** [comment]
It should make it easier to convert IValues accurately to/from Python since we always will know their types. For C++ it will require specifying the type of list/dict on creation, which is something our API already supports.

File: aten/src/ATen/core/ivalue_inl.h [hunk 1]
```diff
@@ -266,14 +270,22 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
...
-  bool completed() {
+  bool completed() const{
     return completed_;
   }
 
   CAFFE2_API friend std::ostream& operator<<(
       std::ostream& out,
       const Future& v);
 
+  TypePtr type() {
```
Comment (suo) [inline aten/src/ATen/core/ivalue_inl.h]: Should be const

File: aten/src/ATen/core/ivalue_inl.h [hunk 2]
```diff
@@ -266,14 +270,22 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
...
+    if (!completed()) {
```
Comment (suo) [inline aten/src/ATen/core/ivalue_inl.h]: This seems a little strange—should the type of a `Future` change depending on its completion state?

File: aten/src/ATen/core/ivalue_inl.h [hunk 3]
```diff
@@ -135,16 +135,19 @@ struct Future;
...
  private:
-   std::vector<IValue> elements_;
+  std::vector<IValue> elements_;
+  mutable std::shared_ptr<TupleType> type_; // lazily computed for unnamed tuples
 
  public:
-  static c10::intrusive_ptr<Tuple> create(std::vector<IValue> elements_, std::shared_ptr<TupleType> type_) {
-    TORCH_INTERNAL_ASSERT(nullptr != type_.get(), "Type cannot be nullptr");
+  // named tuples have additional type information, so we
+  // direclty create them tagged
```
Comment (suo) [inline aten/src/ATen/core/ivalue_inl.h]: spelling: directly

File: aten/src/ATen/core/ivalue.cpp [hunk 4]
```diff
@@ -13,8 +13,63 @@ CAFFE2_API c10::intrusive_ptr<ConstantString> ConstantString::create(
...
+    case Tag::GenericList:
+      return ListType::create(toGenericList().elementType());
+    case Tag::Future:
+      return toFuture()->type();
+    case Tag::Device:
+      return DeviceObjType::get();
+    case Tag::Object:
+      return toObjectRef().type();
+    case Tag::Uninitialized:
+      return AnyType::get();
```
Comment (suo) [inline aten/src/ATen/core/ivalue.cpp]: Aside: what is `Uninitialized` anyway? I'm asking because IValues that are `Any` type should be scrutinized

File: aten/src/ATen/core/jit_type.h [hunk 5]
```diff
@@ -547,7 +571,7 @@ struct CAFFE2_API TensorType : public Type {
         sizes_(tensor.sizes().size()),
         strides_(tensor.sizes().size()),
         requires_grad_(tensor.requires_grad()) {
-          if (!tensor.is_mkldnn()) {
+          if (!tensor.is_mkldnn() && !tensor.is_sparse()) {
```
Comment (suo) [inline aten/src/ATen/core/jit_type.h]: what's the reason for this change?

File: aten/src/ATen/core/jit_type.h [hunk 6]
```diff
@@ -641,6 +665,7 @@ struct CAFFE2_API DictType : public Type {
 
   static DictTypePtr create(TypePtr key, TypePtr value) {
     switch (key->kind()) {
+      case TypeKind::AnyType:
```
Comment (suo) [inline aten/src/ATen/core/jit_type.h]: Any is not really a valid key type for dictionaries, right? Since it may end up being something not in the list below.

I suppose there's not an easy way to represent empty dicts without another `KeyType` cluttering things, so I think it's fine to leave. How are we ensuring that it's not possible to somehow construct a dictionary with an invalid key type?

**Review by suo — APPROVED** [review]
lgtm, some minor comments/questions inline

File: aten/src/ATen/core/jit_type.h (hunk 6 above)
Comment (zdevito) [reply aten/src/ATen/core/jit_type.h]: This check is the at-creation-type check that the dict can't possibly work. There is still a at insertion/lookup-time check that the key is hashable. It is valid to have a Any -> X dictionary. It is just the case that dynamically trying to insert a non-hashable key into it will fail.

File: aten/src/ATen/core/ivalue.cpp (hunk 4 above)
Comment (zdevito) [reply aten/src/ATen/core/ivalue.cpp]: It is introduce when we have early returns, break, and continue. It represents an uninitialized IValue that we know because of invariants introduced in the pass cannot be used. This fact isn't representable in the graph directly so this placeholder is needed. The alternative, representing them with optional values, requires a lot of unwrapping logic around use sites.

File: aten/src/ATen/core/ivalue_inl.h (hunk 1 above)
Comment (zdevito) [reply aten/src/ATen/core/ivalue_inl.h]: as currently written I think `value()` is not const because it holds the mutex. I think it isafe to just have it return `type_` which also avoids the changing type tag behavior.

File: aten/src/ATen/core/jit_type.h (hunk 5 above)
Comment (zdevito) [reply aten/src/ATen/core/jit_type.h]: we can't ask for strides on tensors that do not have them.  We now have tests that ask for the type of tuples, some of which contain sparse tensors, which uncovered this bug.
